### PR TITLE
Add dotnet distroless

### PR DESCRIPTION
**Pull request by salrashid123**
Dotnet distroless build sample.

main points to note:

- bazel build rules for dotnet do not exist yet [issue #39](https://github.com/bazelbuild/rules_dotnet/issues/39) which means for now the dotnet binary needs to get generate outside of of bazel.

- no gcr.io/distroless/dotnet exists yet but the definition for it is contained here.

related:  here is a previous post on plain [dotnet multi-stage builds](https://medium.com/@salmaan.rashid/multi-stage-docker-builds-with-net-ef66fc98b51d) as its related

### CODE REVIEW

**Comment by salrashid123** [comment]
I guess the CI is failing because its looking for the binary in example/dotnet/bin to be there already?

i do know if i generate the binary by hand, atleast the structured tests pass

```bash
$ bazel test examples/dotnet:hello_test
INFO: Found 1 test target...
Target //examples/dotnet:hello_test up-to-date:
  bazel-bin/examples/dotnet/hello_test
INFO: Elapsed time: 4.286s, Critical Path: 4.04s
//examples/dotnet:hello_test                                             PASSED in 4.0s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
```

```bash
$ bazel test examples/dotnet:hello_test --test_verbose_timeout_warnings
INFO: Found 1 test target...
Target //examples/dotnet:hello_test up-to-date:
  bazel-bin/examples/dotnet/hello_test
INFO: Elapsed time: 0.225s, Critical Path: 0.02s
//examples/dotnet:hello_test                                    (cached) PASSED in 4.0s
  WARNING: //examples/dotnet:hello_test: Test execution time (4.0s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="short" or size="small".

Executed 0 out of 1 test: 1 test passes.
```

thoughts on how i can bypass this or any way to fix it?

File: WORKSPACE [hunk 1]
```diff
@@ -56,6 +56,17 @@ dpkg_list(
         "libpython2.7-minimal",
         "python2.7-minimal",
         "libpython2.7-stdlib",
+
+        #dotnet
+        "libcurl3",
+        "libgcc1",
+        "libstdc++6",
```
Comment (mattmoor) [inline WORKSPACE]: libgcc1 and libstdc++6 are dupes with Java.

File: dotnet/BUILD [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        packages["liblzma5"],
+    ],
+    tars = ["@dotnet//:tar"],
+    entrypoint = [
+      #"/opt/dotnet/dotnet",
+     ],
+    symlinks = {
+        "/usr/bin/dotnet": "/opt/dotnet/dotnet",
+    },     
+)
```
Comment (mattmoor) [inline dotnet/BUILD]: Add a newline at EOF

File: dotnet/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+* coreCLR 2.0.0 Runtime and its dependencies.
+
+and
+
+```
+    base = "//base:base",
+    debs = [
+        packages["libcurl3"],
+        packages["libgcc1"],
+        packages["libstdc++6"],
```
Comment (mattmoor) [inline dotnet/README.md]: I think you want `base="//cc:cc"` and you can drop the direct dependency on `libgcc1` and `libstdc++6`

File: examples/dotnet/BUILD [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+load("@bazel_tools//tools/build_defs/pkg:pkg.bzl", "pkg_tar")
+
+docker_build(
+    name = "hello",
+    base = "//dotnet:dotnet",
+    cmd = [
+        "dotnet",
+        "/bin/Release/netcoreapp2.0/hello.dll",
+    ],
+    files = [":bin"],
```
Comment (mattmoor) [inline examples/dotnet/BUILD]: I wouldn't expect this to work unless you check in the `bin/` directory.  Even then, I wouldn't expect adding the directory via `files=[":dirname"]` to work.

File: examples/dotnet/testdata/hello.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+schemaVersion: "1.0.0"
+commandTests:
+  - name: hello
+    command: ['dotnet', '/bin/Release/netcoreapp2.0/hello.dll']
+    expectedOutput: ['Hello World']
```
Comment (mattmoor) [inline examples/dotnet/testdata/hello.yaml]: newline at EOF

File: BUILD.dotnet [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@bazel_tools//tools/build_defs/pkg:pkg.bzl", "pkg_tar")
+
+pkg_tar(
+    name = "tar",
+    files = glob(["**/*"]),
+    package_dir = "/opt/dotnet",
+    strip_prefix = ".",
+)
```
Comment (mattmoor) [inline BUILD.dotnet]: newline at EOF

File: WORKSPACE (hunk 1 above)
Comment (salrashid123) [reply WORKSPACE]: ok

File: dotnet/BUILD (hunk 2 above)
Comment (salrashid123) [reply dotnet/BUILD]: ok

File: dotnet/README.md (hunk 3 above)
Comment (salrashid123) [reply dotnet/README.md]: personally, i preferred the most baseline image to build off of (vs the cc) and manually declaring and adding stuff in but its not a big deal and changed to cc base

File: examples/dotnet/BUILD (hunk 4 above)
Comment (salrashid123) [reply examples/dotnet/BUILD]: I do need the contents of the /bin folder to get referenced into the bazel rules.  The bin folder contains the hello.dll file that is generated out of band.   Once the .dll is generated, that file is available for distroless to import into the image itself.

I'm trying this now create an empty folder with a blank file in it

```
//examples/dotnet/bin/README.md
```
where i'll explain in README.md to first run `dotnet restore` and `dotnet publish -c Release` first.

That way, atleast the CI passes (hopefully)

File: examples/dotnet/testdata/hello.yaml (hunk 5 above)
Comment (salrashid123) [reply examples/dotnet/testdata/hello.yaml]: ok

File: BUILD.dotnet (hunk 6 above)
Comment (salrashid123) [reply BUILD.dotnet]: ok

**Comment by salrashid123** [comment]
ahh...ok, it got passed the build step but failed on the tests...(since the .dll does't even exist).

not sure how to proceed, should (disable the tests, check in the .dll, something else, etc)

```
Status: Downloaded newer image for gcr.io/gcp-runtimes/structure_test:latest
2017/08/28 15:28:57 Running tests for file /cfg/hello.yaml
--- FAIL: TestAll (0.00s)
    --- FAIL: TestAll/Command_Test:_hello (0.00s)
    	structure_test_v1.go:171: Executing: [dotnet /bin/Release/netcoreapp2.0/hello.dll]
    	structure_test_v1.go:188: stderr: Did you mean to run dotnet SDK commands? Please install dotnet SDK from: 
    		  http://go.microsoft.com/fwlink/?LinkID=798306&clcid=0x409
    	structure_test_v1.go:194: Error running command: exit status 145. Continuing.
    	structure_test.go:56: Expected string 'Hello World' not found in output!
    	structure_test_v1.go:261: Test hello exited with incorrect error code! Expected: 0, Actual: 145
	structure_test.go:45: Total tests run: 1
FAIL
```

File: dotnet/README.md (hunk 3 above)
Comment (mattmoor) [reply dotnet/README.md]: FWIW, it is more than just cosmetic since it affects the layering of the final image.  Relying on `//cc` means you'll share those layers of the filesystem  with `/cc`-based images, which includes `//java` and `//python` based images.

**Comment by mattmoor** [comment]
@dlorenc For his thoughts on the right way to document and test this.

**tl;dr** I don't like the idea of breaking `bazel test //...` in a clean repo.

Perhaps we could find a way to document and test this with multi-stage Dockerfile?  This is something we need more broadly (e.g. in each language) regardless, so getting testing and examples in place would be prudent.  IIUC structure testing is a thing outside of Bazel, so I like the idea of sharing the `structure_test` yaml between the Bazel and multi-stage Dockerfile structure test invocations.

Another "meta" option would be to *generate* a Bazel tree and run `bazel test //...` there.  I do this to test `bazel run` in `rules_docker`.  Basically I have a `./testing/e2e.sh` script I run on Travis that runs a bunch of end-to-end tests.

WDYT?

**Comment by r2d4** [comment]
Why can't we depend on fetching the dotnet binary in the WORKSPACE and running `restore` and `release`?  Or we could fetch a prebuilt dll that we host.

**Comment by salrashid123** [comment]
i could try to check those five files  directly into the repo if needed (eg basically have a prebuilt .dll directly in the repo).   We'd do this until there is actual support for bazel+dotnet after which we'll remove the .dll and actually build properly end-to-end.

**Comment by salrashid123** [comment]
ok, i added in the .dll and submitted...the actual dotnet tests pass now but for some reason the ci fails stil

LMK if its ok to add in hello.dll directly in like this (and/or hints on the CI failure above)

```
/examples/dotnet:hello_test                                             PASSED in 6.5s
```

**Comment by r2d4** [comment]
You need to  `Run 'buildifier -mode fix $(find . -name 'BUILD*' -o -name 'WORKSPACE*' -o -name '*.bzl' -type f)' to fix formatting`

Also, do we need the .pdb?

**Comment by salrashid123** [comment]
ci passed.  (didn't know bazel was that particular wrt formatting).

anyway, only these two files are necessary for it to run:
```
hello.dll
hello.runtimeconfig.json
```

so i deleted the others

**Comment by salrashid123** [comment]
hmmm...probably should add those files i just removed back in: its what is generated by default by dotnet so its more 'consistent' (i.,e if i have to regenerate the files from scratch, i don't have to go out of my way to delete the extra files.  

Is there any issue with just keeping all of them in?

```
bin/Release/netcoreapp2.0/hello.dll
bin/Release/netcoreapp2.0/hello.deps.json
bin/Release/netcoreapp2.0/hello.pdb
bin/Release/netcoreapp2.0/hello.runtimeconfig.dev.json
bin/Release/netcoreapp2.0/hello.runtimeconfig.json
```

File: examples/dotnet/bin/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+Please first run
```
Comment (r2d4) [inline examples/dotnet/bin/README.md]: Do you still need to run these if we're checking the files in?

File: dotnet/README.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+which will generate a contained binary:
+
+```
+$ bin/Release/netcoreapp2.0/ubuntu.14.04-x64/publish/hello 
+Hello World
+```
+
+which inturn makes the build rule just
```
Comment (r2d4) [inline dotnet/README.md]: sp: in turn

**Comment by r2d4** [comment]
I think checking in all the generated files is fine.

File: examples/dotnet/bin/README.md (hunk 7 above)
Comment (salrashid123) [reply examples/dotnet/bin/README.md]: those were just notes describing how to these files got there (without them, nobody would really know the specific way we generated these files).  I'll remove this readme.md but add in a snippet about it into the main readme

Comment (r2d4) [reply examples/dotnet/bin/README.md]: SGTM

**Comment by mattmoor** [comment]
@r2d4 We can tell GitHub to squash and merge

**Comment by r2d4** [comment]
Yeah, personal preference but I don't really like the github squash and merge.  I was more suggesting rebasing on master and squashing the typo and review fixes, but possibly keeping some distinct commits (at the author's preference).  

Fine with either
